core/test/factorization/elimination_forest.cpp

```diff
@@ -0,0 +1,203 @@
+{
+    using matrix_type = typename TestFixture::matrix_type;
+    using index_type = typename TestFixture::index_type;
+    auto mtx = gko::initialize<typename TestFixture::matrix_type>(
+        {{1, 0, 1, 0, 0, 0, 0, 1, 0, 0},
+         {0, 1, 0, 1, 0, 0, 0, 0, 0, 1},
+         {1, 0, 1, 0, 0, 0, 0, 0, 0, 0},
+         {0, 0, 0, 1, 0, 0, 0, 0, 1, 1},
+         {0, 1, 0, 0, 1, 0, 0, 0, 1, 1},
+         {0, 0, 0, 0, 0, 1, 0, 1, 0, 0},
+         {0, 0, 1, 0, 0, 1, 1, 0, 0, 0},
+         {1, 0, 0, 0, 0, 1, 0, 1, 1, 1},
+         {0, 0, 0, 1, 1, 0, 0, 1, 1, 0},
+         {0, 1, 0, 1, 1, 0, 0, 1, 0, 1}},
+        this->ref);
+
+    auto forest = gko::factorization::compute_elim_forest(mtx.get());
+
+    GKO_ASSERT_ARRAY_EQ(forest.parents,
+                        I<index_type>({2, 4, 6, 8, 8, 6, 7, 8, 9, 10}));
+    GKO_ASSERT_ARRAY_EQ(forest.child_ptrs,
+                        I<index_type>({0, 0, 0, 1, 1, 2, 2, 4, 5, 8, 9, 10}));
+    GKO_ASSERT_ARRAY_EQ(forest.children,
+                        I<index_type>({0, 1, 2, 5, 6, 3, 4, 7, 8, 9}));
+    GKO_ASSERT_ARRAY_EQ(forest.postorder,
+                        I<index_type>({3, 1, 4, 0, 2, 5, 6, 7, 8, 9}));
+    GKO_ASSERT_ARRAY_EQ(forest.inv_postorder,
+                        I<index_type>({3, 1, 4, 0, 2, 5, 6, 7, 8, 9}));
+    GKO_ASSERT_ARRAY_EQ(forest.postorder_parents,
+                        I<index_type>({8, 2, 8, 4, 6, 6, 7, 8, 9, 10}));
+}
+
+
+TYPED_TEST(EliminationForest, WorksForSeparable)
+{
+    using matrix_type = typename TestFixture::matrix_type;
+    using index_type = typename TestFixture::index_type;
+    auto mtx = gko::initialize<typename TestFixture::matrix_type>(
+        {
+            {1, 0, 1, 0, 0, 0, 0, 0, 0, 0},
+            {0, 1, 1, 0, 0, 0, 0, 0, 0, 0},
+            {1, 1, 1, 0, 0, 0, 0, 0, 0, 0},
+            {0, 0, 0, 1, 1, 0, 0, 0, 0, 0},
+            {0, 0, 0, 1, 1, 1, 0, 0, 0, 1},
+            {0, 0, 0, 0, 1, 1, 0, 0, 0, 0},
+            {0, 0, 0, 0, 0, 0, 1, 0, 0, 1},
+            {0, 0, 0, 0, 0, 0, 1, 1, 0, 0},
+            {0, 0, 0, 0, 0, 0, 0, 0, 1, 1},
+            {0, 0, 0, 0, 1, 0, 1, 0, 1, 1},
+        },
+        this->ref);
+
+    auto forest = gko::factorization::compute_elim_forest(mtx.get());
+
+    GKO_ASSERT_ARRAY_EQ(forest.parents,
+                        I<index_type>({2, 2, 10, 4, 5, 9, 7, 9, 9, 10}));
+    GKO_ASSERT_ARRAY_EQ(forest.child_ptrs,
+                        I<index_type>({0, 0, 0, 2, 2, 3, 4, 4, 5, 5, 8, 10}));
+    GKO_ASSERT_ARRAY_EQ(forest.children,
+                        I<index_type>({0, 1, 3, 4, 6, 5, 7, 8, 2, 9}));
+    GKO_ASSERT_ARRAY_EQ(forest.postorder,
+                        I<index_type>({0, 1, 2, 3, 4, 5, 6, 7, 8, 9}));
+    GKO_ASSERT_ARRAY_EQ(forest.inv_postorder,
+                        I<index_type>({0, 1, 2, 3, 4, 5, 6, 7, 8, 9}));
+    GKO_ASSERT_ARRAY_EQ(forest.postorder_parents,
+                        I<index_type>({2, 2, 10, 4, 5, 9, 7, 9, 9, 10}));
+}
+
+
```

Comment:
In the test cases, the postorder array is self-inverted for some reason. I'd suggest adding this test where this is not the case:

```suggestion

TYPED_TEST(EliminationForest, WorksForPostOrderNotSelfInverted)
{
    using matrix_type = typename TestFixture::matrix_type;
    using index_type = typename TestFixture::index_type;
    auto mtx = gko::initialize<typename TestFixture::matrix_type>(
        {
            {1, 0, 1, 0, 0, 0, 0, 0, 0, 0},
            {0, 1, 0, 0, 1, 0, 0, 0, 0, 0},
            {1, 0, 1, 0, 0, 0, 1, 0, 0, 0},
            {0, 0, 0, 1, 0, 0, 0, 0, 1, 0},
            {0, 1, 0, 0, 1, 1, 0, 0, 0, 0},
            {0, 0, 0, 0, 1, 1, 1, 0, 0, 0},
            {0, 0, 1, 0, 0, 1, 1, 1, 0, 0},
            {0, 0, 0, 0, 0, 0, 1, 1, 1, 0},
            {0, 0, 0, 1, 0, 0, 0, 1, 1, 1},
            {0, 0, 0, 0, 0, 0, 1, 0, 1, 1},
        },
        this->ref);

    auto forest = gko::factorization::compute_elim_forest(mtx.get());

    GKO_ASSERT_ARRAY_EQ(forest.parents,
                        I<index_type>({2, 4, 6, 8, 5, 6, 7, 8, 9, 10}));
    GKO_ASSERT_ARRAY_EQ(forest.child_ptrs,
                        I<index_type>({0, 0, 0, 1, 1, 2, 3, 5, 6, 8, 9, 10}));
    GKO_ASSERT_ARRAY_EQ(forest.children,
                        I<index_type>({0, 1, 4, 2, 5, 6, 3, 7, 8, 9}));
    GKO_ASSERT_ARRAY_EQ(forest.postorder,
                        I<index_type>({3, 0, 2, 1, 4, 5, 6, 7, 8, 9}));
    GKO_ASSERT_ARRAY_EQ(forest.inv_postorder,
                        I<index_type>({1, 3, 2, 0, 4, 5, 6, 7, 8, 9}));
    GKO_ASSERT_ARRAY_EQ(forest.postorder_parents,
                        I<index_type>({8, 2, 6, 4, 5, 6, 7, 8, 9, 10}));
}


```